x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/IndicesPermission.java

```diff
@@ -93,34 +122,51 @@ public IndicesPermission(Group... groups) {
         return indices::contains;
     }
 
-    private static Predicate<String> buildAutomataPredicate(List<String> indices) {
+    private static Predicate<String> buildAutomataPredicate(final Collection<String> indices) {
+        final Predicate<String> indicesPredicate;
         try {
-            return Automatons.predicate(indices);
+            indicesPredicate = Automatons.predicate(indices);
         } catch (TooComplexToDeterminizeException e) {
-            LogManager.getLogger(IndicesPermission.class).debug("Index pattern automaton [{}] is too complex", indices);
+            logger.debug("Index pattern automaton [{}] is too complex", indices);
             String description = Strings.collectionToCommaDelimitedString(indices);
             if (description.length() > 80) {
                 description = Strings.cleanTruncate(description, 80) + "...";
             }
             throw new ElasticsearchSecurityException("The set of permitted index patterns [{}] is too complex to evaluate", e, description);
         }
-    }
-
-    @Override
-    public Iterator<Group> iterator() {
-        return Arrays.asList(groups).iterator();
+        return (index) -> {
+            assert false == isIndexPattern(index);
+            if (indicesPredicate.test(index)) {
+                if (isSystemIndex(index)) {
+                    logger.debug("Index pattern automaton [{}] cannot match system indices [{}]", indices, SystemIndicesNames.indexNames());
+                    return false;
+                } else {
+                    return true;
+                }
+            } else {
+                return false;
+            }
+        };
```

Comment:
The predicate is splayed out like this to accommodate for the logging statement.